File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/autos/BlueLeftDetection.java
package org.firstinspires.ftc.teamcode.autos;

import com.qualcomm.robotcore.eventloop.opmode.Autonomous;

import org.firstinspires.ftc.teamcode.Hardware;
import org.firstinspires.ftc.teamcode.PosDetector;

@Autonomous(group = "testing")
public class BlueLeftDetection extends DetectionAuto {
    @Override
    protected void preStart() throws InterruptedException {
        hw.setGrabberIsOpen(false);
    }

    @Override
    protected void autoCode() throws InterruptedException {
        int position = detectedPositionNow();
        int level = position == 0 ? 1 : 2; // 0 is leftmost is bottom level
        telemetry.addData("level", level);
        telemetry.update();

        hw.resetGrabberEncoder();
        hw.driveInches(24);
        hw.setGrabberHeight(level == 1 ? Hardware.GrabberHeight.FIRST_LEVEL : Hardware.GrabberHeight.SECOND_LEVEL);
        hw.goToHeading(45);
        hw.driveInches(5);
        hw.setGrabberIsOpen(true);
        hw.driveInches(-5);
        hw.goToHeading(-90);
    }

    @Override
    protected PosDetector getPosDetector() {
        return PosDetectorFactory.blueLeft(telemetry);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/autos/BlueLeftDetection.kt
package org.firstinspires.ftc.teamcode.autos

import com.qualcomm.robotcore.eventloop.opmode.Autonomous
import org.firstinspires.ftc.teamcode.utils.Hardware
import org.firstinspires.ftc.teamcode.utils.PosDetector
import org.firstinspires.ftc.teamcode.utils.Side

@Autonomous(preselectTeleOp = "Blue Tank Drive")
class BlueLeftDetection : DetectionAuto() {
    override fun getNewPosDetector(): PosDetector = PosDetectorFactory.blueLeft(telemetry)

    override fun beforeStart() {
        super.beforeStart()
        telemetry.clearAll()
    }

    override fun mainOpMode() {
        hw.grabberIsOpen = false

        val level = when (detectedPositionNow()) {
            0 -> Hardware.GrabberHeight.THIRD_LEVEL
            1 -> Hardware.GrabberHeight.SECOND_LEVEL
            else -> Hardware.GrabberHeight.FIRST_LEVEL
        }
        telemetry.printData("level", level)

        hw.resetGrabberEncoder()

        // deposit preload
        hw.driveInches(20.0)
        if (level == Hardware.GrabberHeight.THIRD_LEVEL) {
            hw.driveInches(20.0)
            hw.driveInches(-20.0)
            hw.turnToHeading(45.0)
            hw.grabberHeight = level
        } else {
            hw.turnToHeading(45.0)
            hw.grabberHeight = level
        }

        hw.driveInches(7.0)
        hw.grabberIsOpen = true

        //park
        hw.driveInches(-7.0)
        hw.turnToHeading(0.0)
        hw.driveInches(-10.0)
        hw.grabberHeight = Hardware.GrabberHeight.FIRST_LEVEL
        hw.turnToHeading(-90.0)
        hw.driveInches(45.0)
    }
}